cache/blobs.go

```diff
@@ -468,24 +481,36 @@ func ensureCompression(ctx context.Context, ref *immutableRef, comp compression.
 
 		// First, lookup local content store
 		if _, err := ref.getBlobWithCompression(ctx, comp.Type); err == nil {
-			return nil, nil // found the compression variant. no need to convert.
+			return nil, nil // found the compression variant. no need to ensure compression.
 		}
 
-		// Convert layer compression type
-		if err := (lazyRefProvider{
-			ref:     ref,
-			desc:    desc,
-			dh:      ref.descHandlers[desc.Digest],
-			session: s,
-		}).Unlazy(ctx); err != nil {
-			return nil, err
-		}
-		newDesc, err := layerConvertFunc(ctx, ref.cm.ContentStore, desc)
-		if err != nil {
-			return nil, errors.Wrapf(err, "failed to convert")
+		var newDesc *ocispecs.Descriptor
+		if errors.Is(cvtErr, errdefs.ErrNotImplemented) {
+			defer ref.cm.mu.Unlock()
+			ref.cm.mu.Lock()
```

Comment:
Why does the cache manager need to be locked for the rest of the function in this branch?

Reply:
Thanks for the reply, fixed by removing the big lock. The background is since concurrently calling computeBlob to create `sr.cm.ContentStore.Writer(...)` when converting nydus blob to multiple different compression types at the same time will throw an error that ref lock is unavailable. The new fix has been updated [here](https://github.com/moby/buildkit/pull/2581/commits/0084eb3c49c9e6722d6790ef282e4816e4158240#diff-57fef1d2d10523a9c84ae35b0c4c5aa6e53016ace8c20facf3e6853df7c0d632R165) .